File: backend/src/SmartPayEngine.Core/Services/SmartRoutingService.cs
using SmartPayEngine.Core.Abstractions;
using SmartPayEngine.Core.Catalogs;
using SmartPayEngine.Core.Common;
using SmartPayEngine.Core.Entities;
using SmartPayEngine.Core.Enums;
using SmartPayEngine.Core.Exceptions;
using SmartPayEngine.Core.Models;
using SmartPayEngine.Core.ValueObjects;

namespace SmartPayEngine.Core.Services;

/// <summary>
/// Scores every eligible payment rail for a transaction across four dimensions
/// (cost, speed, reliability, reversibility) and returns a ranked list plus the
/// winner. The weighting is policy-driven so callers can optimize for cost,
/// speed, or a balanced blend.
/// </summary>
public sealed class SmartRoutingService : ISmartRoutingService
{
    /// <summary>Number of options returned (top-N after ranking).</summary>
    private const int MaxOptions = 5;

    /// <summary>Settlement time (hours) at which the speed score bottoms out.</summary>
    private const double SlowestSettlementHours = 96;

    public RoutingRecommendation Analyze(Transaction transaction, RoutingPolicy? policy = null)
        => Route(transaction.Amount, transaction.ToCurrency, urgent: false, policy);

    public RoutingRecommendation Analyze(
        Money amount,
        Currency toCurrency,
        bool urgent = false,
        RoutingPolicy? policy = null)
        => Route(amount, toCurrency, urgent, policy);

    private RoutingRecommendation Route(
        Money amount,
        Currency toCurrency,
        bool urgent,
        RoutingPolicy? policy)
    {
        if (amount.IsZero)
        {
            throw new DomainException("Routing requires a positive amount.", nameof(amount));
        }

        var resolvedPolicy = policy ?? RoutingPolicy.Balanced;
        var routes = PaymentRoute.AvailableFor(amount.Currency, toCurrency);
        if (routes.Count == 0)
        {
            throw new DomainException(
                $"No payment rail supports {amount.Currency} → {toCurrency}.",
                nameof(toCurrency));
        }

        // Pre-compute the fee range so the cost dimension can be normalized
        // (cheapest route scores 1, dearest scores 0).
        var fees = routes.ToDictionary(r => r.Provider, r => FeeFor(r.Profile, amount));
        var cheapest = new Money(fees.Values.Min(f => f.Amount), amount.Currency);
        var dearest = new Money(fees.Values.Max(f => f.Amount), amount.Currency);

        var options = routes
            .Select(r => Evaluate(r.Profile, amount, fees[r.Provider], cheapest, dearest, urgent, resolvedPolicy))
            .OrderByDescending(o => o.Score)
            .Take(MaxOptions)
            .ToList();

        return new RoutingRecommendation(options[0], options, resolvedPolicy);
    }

    /// <summary>Total fee = variable (bps) + flat, in the amount's currency.</summary>
    private static Money FeeFor(ProviderProfile profile, Money amount)
    {
        var variable = amount.Amount * profile.VariableCostBps / 10_000m;
        return new Money(variable + profile.FlatFee, amount.Currency);
    }

    private static RoutingOption Evaluate(
        ProviderProfile profile,
        Money amount,
        Money fee,
        Money cheapest,
        Money dearest,
        bool urgent,
        RoutingPolicy policy)
    {
        // Cost: cheapest = 1, dearest = 0.
        var costScore = dearest.Amount == cheapest.Amount
            ? 1.0
            : 1.0 - Scoring.Lerp((double)fee.Amount, (double)cheapest.Amount, (double)dearest.Amount, 0, 1);

        // Speed: instant = 1, decaying to 0 by ~96h. Urgency sharpens the curve.
        var speedBase = 1.0 - Scoring.Lerp(profile.SettlementTime.TotalHours, 0, SlowestSettlementHours, 0, 1);
        var speedScore = urgent ? Math.Sqrt(speedBase) : speedBase;

        var reliabilityScore = Scoring.Clamp(profile.Reliability, 0, 1);
        var reversibilityScore = Scoring.Clamp(profile.BaseReversalSuccess, 0, 1);

        var score = Scoring.WeightedAverage(new[]
        {
            new ScoreFactor("cost", costScore, policy.Cost, string.Empty),
            new ScoreFactor("speed", speedScore, policy.Speed, string.Empty),
            new ScoreFactor("reliability", reliabilityScore, policy.Reliability, string.Empty),
            new ScoreFactor("reversibility", reversibilityScore, policy.Reversibility, string.Empty)
        });

        var rationale =
            $"{profile.DisplayName}: fee {fee}, ~{FormatDuration(profile.SettlementTime)} to settle, " +
            $"{profile.Reliability:P0} reliability, {profile.Reversibility} reversibility.";

        return new RoutingOption(
            profile.Provider,
            profile.DisplayName,
            fee,
            profile.SettlementTime,
            reliabilityScore,
            profile.Reversibility,
            Math.Round(score, 4),
            rationale);
    }

    private static string FormatDuration(TimeSpan span)
        => span == TimeSpan.Zero ? "instant" : $"{span.TotalHours:0}h";
}

File: src/domain/services/smart-routing.service.ts
/**
 * `SmartRoutingService` — evaluates the candidate payment rails for a given
 * payment intent and recommends the best one.
 *
 * "Best" is policy-driven: the caller supplies an {@link RoutingPolicy} that
 * weights cost, speed, reliability and reversibility. The service scores every
 * eligible rail on each dimension, combines them into a weighted score, and
 * returns a ranked list plus the winner. This is the same rail-economics model
 * the reversal engine consumes, kept in one place.
 */
import type { Currency } from "@/lib/money";
import { Money } from "../money";
import { clamp, lerp, weightedAverage } from "../shared/math";
import { DomainErrorCode, err, ok, type Result } from "../shared/result";
import { PaymentRail, railsForCurrency, type RailProfile } from "../entities/payment-rail";

/** What the caller is trying to optimize for. Weights need not sum to 1. */
export interface RoutingPolicy {
  readonly weightCost: number;
  readonly weightSpeed: number;
  readonly weightReliability: number;
  /** How much to value the ability to reverse the payment later. */
  readonly weightReversibility: number;
  /** If true, rails that can't carry the amount/currency are excluded. */
  readonly requireReversible?: boolean;
}

/** A sensible default that balances all four dimensions. */
export const BALANCED_POLICY: RoutingPolicy = {
  weightCost: 0.3,
  weightSpeed: 0.25,
  weightReliability: 0.25,
  weightReversibility: 0.2,
};

export interface PaymentIntent {
  readonly amount: Money;
  readonly currency: Currency;
  /** True when the customer needs the funds to arrive quickly. */
  readonly urgent?: boolean;
}

/** Per-rail scoring breakdown returned to the caller for transparency. */
export interface RouteEvaluation {
  readonly rail: PaymentRail;
  readonly profile: RailProfile;
  /** Total fee for this rail given the amount, in minor units. */
  readonly estimatedFeeMinor: number;
  /** Normalized sub-scores in [0, 1]. */
  readonly costScore: number;
  readonly speedScore: number;
  readonly reliabilityScore: number;
  readonly reversibilityScore: number;
  /** Final weighted score in [0, 1]; higher is better. */
  readonly score: number;
}

export interface RoutingDecision {
  readonly recommended: RouteEvaluation;
  /** All eligible rails, ranked best-first. */
  readonly ranked: readonly RouteEvaluation[];
  readonly policy: RoutingPolicy;
}

export class SmartRoutingService {
  constructor(private readonly policy: RoutingPolicy = BALANCED_POLICY) {}

  /** Total cost (variable bps + flat fee) for a rail at this amount. */
  private feeMinor(profile: RailProfile, amount: Money): number {
    const variable = Math.round((amount.minor * profile.costBps) / 10_000);
    return variable + profile.flatFeeMinor;
  }

  /**
   * Score a single rail against the intent. All sub-scores are normalized to
   * [0, 1] so the policy weights are comparable.
   */
  private evaluateRail(
    profile: RailProfile,
    intent: PaymentIntent,
    cheapestFee: number,
    dearestFee: number,
  ): RouteEvaluation {
    const fee = this.feeMinor(profile, intent.amount);

    // Cost: cheapest rail scores 1, dearest scores 0.
    const costScore = dearestFee === cheapestFee ? 1 : 1 - lerp(fee, cheapestFee, dearestFee, 0, 1);

    // Speed: instant = 1, decaying toward 0 by ~96h. Urgency sharpens it.
    const speedBase = 1 - lerp(profile.settlementHours, 0, 96, 0, 1);
    const speedScore = intent.urgent ? speedBase ** 0.5 : speedBase;

    const reliabilityScore = clamp(profile.reliability, 0, 1);
    const reversibilityScore = clamp(profile.baseReversalSuccess, 0, 1);

    const score = weightedAverage([
      { label: "cost", score: costScore, weight: this.policy.weightCost },
      { label: "speed", score: speedScore, weight: this.policy.weightSpeed },
      {
        label: "reliability",
        score: reliabilityScore,
        weight: this.policy.weightReliability,
      },
      {
        label: "reversibility",
        score: reversibilityScore,
        weight: this.policy.weightReversibility,
      },
    ]);

    return {
      rail: profile.rail,
      profile,
      estimatedFeeMinor: fee,
      costScore,
      speedScore,
      reliabilityScore,
      reversibilityScore,
      score,
    };
  }

  /**
   * Evaluate all eligible rails for the intent and pick a winner.
   */
  route(intent: PaymentIntent): Result<RoutingDecision> {
    if (!intent.amount.isPositive) {
      return err(DomainErrorCode.Validation, "Routing requires a positive amount", "amount");
    }

    let candidates = railsForCurrency(intent.currency);
    if (this.policy.requireReversible) {
      candidates = candidates.filter((p) => p.baseReversalSuccess >= 0.5);
    }
    if (candidates.length === 0) {
      return err(
        DomainErrorCode.Unsupported,
        `No payment rail available for ${intent.currency} under the current policy`,
        "currency",
      );
    }

    // Pre-compute fee range so the cost dimension can be normalized.
    const fees = candidates.map((p) => this.feeMinor(p, intent.amount));
    const cheapest = Math.min(...fees);
    const dearest = Math.max(...fees);

    const ranked = candidates
      .map((p) => this.evaluateRail(p, intent, cheapest, dearest))
      .sort((a, b) => b.score - a.score);

    return ok({ recommended: ranked[0], ranked, policy: this.policy });
  }
}
